Gazel-Tutorial/src/Gazel.Tutorial.Module.ProductManagement/Product.cs

```diff
@@ -0,0 +1,97 @@
+﻿using Gazel;
+using Gazel.DataAccess;
+using Gazel.Tutorial.Module.ProductManagement.Service;
+
+namespace Gazel.Tutorial.Module.ProductManagement
+{
+
+    public class Product : IProductInfo, IProductService
+    {
+        private IRepository<Product> repository;
+
+        protected Product() { }
+
+        public Product(IRepository<Product> repository)
+        {
+            this.repository = repository;
+        }
+
+        public virtual int Id { get; protected set; }
+        public virtual string ProductName { get; protected set; }
+        public virtual float Price { get; protected set; }
+        public virtual int Stock { get; protected set; }
+
+        protected internal virtual Product With(string name, float price, int stock)
+        {
+            ProductName = name;
+            Price = price;
+            Stock = stock;
+
+            repository.Insert(this);
+
+            return this;
+        }
+        public virtual void UpdateProduct(string name = null, float price = default(float), int stock = default(int))
+        {
+            if (name.IsNullOrWhiteSpace()) name = ProductName;
+            if (price.IsDefault()) price = Price;
+            if (stock.IsDefault()) stock = Stock;
+
+            ProductName = name;
+            Price = price;
+            Stock = stock;
+        }
+
+        public virtual void RemoveProduct()
+        {
+            repository.Delete(this);
+        }
+    }
+
+    public class Products : Query<Product>, IProductsService
+    {
+        public Products(IModuleContext context) : base(context) { }
+
+        public List<Product> ByName(string name)
+        {
+            return By(t => t.ProductName == name);
+        }
+
+        public List<Product> ByPositiveStock()
+        {
+            return By(t => t.Stock > 0);
+        }
+
+        public List<Product> ByPriceHigherThan(float lowerBound)
+        {
+            return By(t => t.Price >= lowerBound);
+        }
+
+        public List<Product> ByPriceLowerThan(float upperBound)
+        {
+            return By(t => t.Price <= upperBound);
+        }
+
+        public List<Product> ByPriceRange(float lowerBound, float upperBound)
+        {
+            return ByPriceHigherThan(lowerBound).Intersect(ByPriceLowerThan(upperBound)).ToList();
+        }
+
+        public Product SingleById(int productId)
+        {
+            return SingleBy(t => t.Id == productId);
+        }
+
+        IProductInfo IProductsService.GetProduct(int productId) =>
```

Comment:
SingleById'ler bundan sanirim, burada direkt Product product alabilirsin parametre olarak

http://gazel.io/conventions/#persistent-objects-as-parameters

Reply:
product için olan SingleById bunun içinmiş, int productId'yi Product product yaptım.